File: 29968040/python-57482691/minecraft/particle/ParticleEngine.py
from .Particle import Particle
from minecraft.world.Tesselator import Tesselator
from minecraft.Textures import Textures

import math
import GL11


# 粒子引擎,渲染破坏粒子效果
class ParticleEngine:
    def __init__(self, world):
        self.world = world
        self.particles = []

    '''
     * 添加粒子
     * @param p
    '''
    def add(self, p: Particle):
        self.particles.append(p)

    def tick(self):
        i = 0
        while i < len(self.particles):
            p: Particle = self.particles[i]
            p.tick()
            if p.removed:
                del self.particles[i]
                i -= 1
            i += 1

    def render(self, player, a: float, layer: int):
        GL11.glEnable(GL11.GL_TEXTURE_2D)
        _id: int = Textures.loadTexture("/terrain.png", GL11.GL_NEAREST)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, _id)
        xa: float = -math.cos(player.yRot * math.pi / 180.0)
        za: float = -math.sin(player.yRot * math.pi / 180.0)

        xa2: float = -za * math.sin(player.xRot * math.pi / 180.0)
        za2: float = xa * math.sin(player.xRot * math.pi / 180.0)
        ya: float = math.cos(player.xRot * math.pi / 180.0)

        t: Tesselator = Tesselator.instance
        GL11.glColor4f(0.8, 0.8, 0.8, 1.0)
        t.init()
        for i in range(len(self.particles)):
            p: Particle = self.particles[i]
            if p.isLit() ^ layer == 1:
                p.render(t, a, xa, ya, za, xa2, za2)
        t.flush()
        GL11.glDisable(GL11.GL_TEXTURE_2D)

File: 29968040/python-57482691/minecraft/particle/ParticleEngine.java
package com.mojang.minecraft.particle;

import com.mojang.minecraft.Player;
import com.mojang.minecraft.Textures;
import com.mojang.minecraft.world.World;
import com.mojang.minecraft.world.Tesselator;
import java.util.ArrayList;
import java.util.List;
import org.lwjgl.opengl.GL11;
/*
 * 粒子引擎,渲染破坏粒子效果
 */
public class ParticleEngine {
	protected World world;
	private List<Particle> particles = new ArrayList();

	public ParticleEngine(World world) {
		this.world = world;
	}

	/**
	 * 添加粒子
	 * @param p
	 */
	public void add(Particle p) {
		this.particles.add(p);
	}

	public void tick() {
		for (int i = 0; i < this.particles.size(); i++) {
			Particle p = (Particle) this.particles.get(i);
			p.tick();
			if (p.removed) {
				this.particles.remove(i--);
			}
		}
	}

	public void render(Player player, float a, int layer) {
		GL11.glEnable(3553);
		int id = Textures.loadTexture("/terrain.png", 9728);
		GL11.glBindTexture(3553, id);
		float xa = -(float) Math.cos(player.yRot * 3.141592653589793D / 180.0D);
		float za = -(float) Math.sin(player.yRot * 3.141592653589793D / 180.0D);

		float xa2 = -za
				* (float) Math.sin(player.xRot * 3.141592653589793D / 180.0D);
		float za2 = xa
				* (float) Math.sin(player.xRot * 3.141592653589793D / 180.0D);
		float ya = (float) Math.cos(player.xRot * 3.141592653589793D / 180.0D);

		Tesselator t = Tesselator.instance;
		GL11.glColor4f(0.8F, 0.8F, 0.8F, 1.0F);
		t.init();
		for (int i = 0; i < this.particles.size(); i++) {
			Particle p = (Particle) this.particles.get(i);
			if ((p.isLit() ^ layer == 1)) {
				p.render(t, a, xa, ya, za, xa2, za2);
			}
		}
		t.flush();
		GL11.glDisable(3553);
	}
}

/*
 * Location:
 * H:\Game\Minecraft1.12.2\Minecraft1.12.2\.minecraft\versions\rd-161348
 * \rd-161348.jar Qualified Name: com.mojang.minecraft.particle.ParticleEngine
 * JD-Core Version: 0.6.2
 */